wcfsetup/install/files/lib/system/user/authentication/password/algorithm/TPhpass.class.php

```diff
@@ -0,0 +1,112 @@
+<?php
+namespace wcf\system\user\authentication\password\algorithm;
+
+/**
+ * Implementation of the PHPASS password algorithm.
+ *
+ * @author	Joshua Ruesweg
+ * @copyright	2001-2020 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\System\User\Authentication\Password\Algorithm
+ * @since	5.4
+ */
+trait TPhpass {
+	private $itoa64 = './0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
+	
+	/**
+	 * Returns the hashed password, with the given settings.
+	 */
+	private function hashPhpass(string $password, string $settings): string {
+		$output = '*';
+		
+		// Check for correct hash
+		if (substr($settings, 0, 3) !== '$H$' && substr($settings, 0, 3) !== '$P$') {
+			return $output;
+		}
+		
+		$count_log2 = strpos($this->itoa64, $settings[3]);
+		
+		if ($count_log2 < 7 || $count_log2 > 30) {
+			return $output;
+		}
+		
+		$count = 1 << $count_log2;
+		$salt = substr($settings, 4, 8);
+		
+		if (strlen($salt) != 8) {
+			return $output;
+		}
+		
+		$hash = md5($salt . $password, true);
+		do {
+			$hash = md5($hash . $password, true);
+		}
+		while (--$count);
+		
+		$output = substr($settings, 0, 12);
+		$hash_encode64 = function ($input, $count, &$itoa64) {
+			$output = '';
+			$i = 0;
+			
+			do {
+				$value = ord($input[$i++]);
+				$output .= $itoa64[$value & 0x3f];
+				
+				if ($i < $count) {
+					$value |= ord($input[$i]) << 8;
+				}
+				
+				$output .= $itoa64[($value >> 6) & 0x3f];
+				
+				if ($i++ >= $count) {
+					break;
+				}
+				
+				if ($i < $count) {
+					$value |= ord($input[$i]) << 16;
+				}
+				
+				$output .= $itoa64[($value >> 12) & 0x3f];
+				
+				if ($i++ >= $count) {
+					break;
+				}
+				
+				$output .= $itoa64[($value >> 18) & 0x3f];
+			}
+			while ($i < $count);
+			
+			return $output;
+		};
+		
+		$output .= $hash_encode64($hash, 16, $this->itoa64);
+		
+		return $output;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function verify(string $password, string $hash): bool {
+		[$hash, $salt] = explode(':', $hash, 2);
+		
+		return \hash_equals($hash, $this->hashPhpass($password, $hash));
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function hash(string $password): string {
+		$settings = '$H$8';
+		$settings .= \bin2hex(\random_bytes(4));
+		
+		return $this->hashPhpass($password, $settings).':';
```

Comment:
```suggestion
		return $this->hashPhpass($password, $settings).':'.$settings;
```
?

Reply:
No, the salt is prefixed together with other parameters during the process.

Reply:
Ah. I misread the `verify()`. Then I am a bit confused about the need for the trailing colon and the `explode` in `verify()`. Is that for backwards compatibility? If yes: Add a comment please.

Reply:
The same way it is exported in the Exporter. I assume that each password is guaranteed to have a hash separated with a `:`. Otherwise the script would throw an error.